main/external/fsharpbinding/MonoDevelop.FSharpBinding/VS/CommonRoslynHelpers.fs

```
@@ -0,0 +1,51 @@
+
+namespace MonoDevelop.FSharp.Editor
+
+open System.Threading.Tasks
+open Microsoft.CodeAnalysis.Text
+open Microsoft.FSharp.Compiler.Range
+
+module CommonRoslynHelpers =
+
+    let FSharpRangeToTextSpan(sourceText: SourceText, range: range) =
+        // Roslyn TextLineCollection is zero-based, F# range lines are one-based
+        let startPosition = sourceText.Lines.[range.StartLine - 1].Start + range.StartColumn
+        let endPosition = sourceText.Lines.[range.EndLine - 1].Start + range.EndColumn
+        TextSpan(startPosition, endPosition - startPosition)
+
+    let TryFSharpRangeToTextSpan(sourceText: SourceText, range: range) : TextSpan option =
+        try Some(FSharpRangeToTextSpan(sourceText, range))
+        with e -> 
+            MonoDevelop.Core.LoggingService.LogInternalError e
+            None
+
+    let TextSpanToFSharpRange(fileName: string, textSpan: TextSpan, sourceText: SourceText) : range =
+        let startLine = sourceText.Lines.GetLineFromPosition textSpan.Start
+        let endLine = sourceText.Lines.GetLineFromPosition textSpan.End
+        mkRange 
+            fileName 
+            (Pos.fromZ startLine.LineNumber (textSpan.Start - startLine.Start))
+            (Pos.fromZ endLine.LineNumber (textSpan.End - endLine.Start))
+
+    let GetCompletedTaskResult(task: Task<'TResult>) =
+        if task.Status = TaskStatus.RanToCompletion then
+            task.Result
+        else
+            raise(task.Exception.GetBaseException())
+
+
+    let StartAsyncAsTask cancellationToken computation =
+        let computation =
+            async {
+                try
+                    return! computation
+                with e ->
+                    //Assert.Exception(e)
+                    MonoDevelop.Core.LoggingService.LogInternalError e
+                    return Unchecked.defaultof<_>
```

wouldn't it be better to raise the exception here instead of returning default value? I know that `return reraise` doesn't work in F# yet (see https://github.com/fsharp/fslang-suggestions/issues/660 ) but the workaround I usually use is `(ExceptionDispatchInfo.Capture ex).Throw ()`

---

See this code [here](https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/Common/RoslynHelpers.fs#L95-L127) where I originally cribbed this code from :)

The comment from this code is clear that the exception should be swallowed. I'm not sure why we are not using the same implementation of StartAsyncAsTask here - some of the code in this PR is taken from a spike I worked on ages ago. I think this is code from the spike where I was just trying to make things work.